sys/kern/proc.c

```diff
@@ -29,26 +38,43 @@ static proc_list_t proc_list = TAILQ_HEAD_INITIALIZER(proc_list);
 static proc_list_t zombie_list = TAILQ_HEAD_INITIALIZER(zombie_list);
 static pgrp_list_t pgrp_list = TAILQ_HEAD_INITIALIZER(pgrp_list);
 
-/* Pid 0 is never available, because of its special treatment by some
- * syscalls e.g. kill. */
-static bitstr_t pid_used[bitstr_size(NPROC)] = {1};
+static pgrp_t *pgrp_lookup(pgid_t pgid);
+static proc_t *proc_find_raw(pid_t pgid);
 
 /* Process ID management functions */
-static pid_t pid_alloc(void) {
-  assert(mtx_owned(all_proc_mtx));
 
-  pid_t pid;
-  bit_ffc(pid_used, NPROC, &pid);
-  if (pid < 0)
-    panic("Out of PIDs!");
-  bit_set(pid_used, pid);
-  return pid;
+static void proc_init_hashtbl(void) {
+  for (int i = 0; i < NBUCKETS; i++) {
+    TAILQ_INIT(&proc_hashtbl[i]);
+    TAILQ_INIT(&pgrp_hashtbl[i]);
+  }
+}
+
+static bool pid_is_taken(pid_t pid) {
+  /* PID 0 is reserved. */
+  if (pid == 0)
+    return true;
+  if (proc_find_raw(pid) != NULL)
```

Comment:
Is `!= NULL` necessary here and below?

Reply:
Semantically, `p` and `p != NULL` are equivalent here. It's a matter of taste and clarity, I suppose.